torch/csrc/Generator.cpp

```diff
@@ -116,37 +116,60 @@ static PyObject* THPGenerator_setState(PyObject* _self, PyObject* _new_state) {
   END_HANDLE_TH_ERRORS
 }
 
-static PyObject* THPGenerator_manualSeed(PyObject* _self, PyObject* seed) {
-  HANDLE_TH_ERRORS
-  auto self = (THPGenerator*)_self;
-  auto generator = self->cdata;
-  THPUtils_assert(
-      THPUtils_checkLong(seed),
-      "manual_seed expected a long, "
-      "but got %s",
-      THPUtils_typename(seed));
-  // See Note [Acquire lock when using random generators]
-  std::lock_guard<std::mutex> lock(generator.mutex());
+uint64_t unpack_int_pyobj(PyObject* obj) {
   // NOLINTNEXTLINE(cppcoreguidelines-init-variables)
-  uint64_t seed_unpacked;
+  uint64_t obj_unpacked;
   try {
     // First try to interpret as unsigned long
-    seed_unpacked = THPUtils_unpackUInt64(seed);
+    obj_unpacked = THPUtils_unpackUInt64(obj);
   } catch (...) {
     if (PyErr_ExceptionMatches(PyExc_OverflowError)) {
-      // If an overflow happened, then the seed could be negative,
+      // If an overflow happened, then the obj could be negative,
       // so try to interpret it as signed long
       PyErr_Clear();
-      int64_t seed_unpacked_signed = THPUtils_unpackLong(seed);
-      seed_unpacked = *(reinterpret_cast<uint64_t*>(&seed_unpacked_signed));
+      int64_t obj_unpacked_signed = THPUtils_unpackLong(obj);
+      obj_unpacked = *(reinterpret_cast<uint64_t*>(&obj_unpacked_signed));
     } else {
       // If any other type of exception happened, rethrow it
       throw;
     }
   }
+  return obj_unpacked;
+}
+
+static PyObject* THPGenerator_manualSeed(PyObject* _self, PyObject* seed) {
+  HANDLE_TH_ERRORS
+  auto self = (THPGenerator*)_self;
+  auto generator = self->cdata;
+  THPUtils_assert(
+      THPUtils_checkLong(seed),
+      "manual_seed expected a long, "
+      "but got %s",
+      THPUtils_typename(seed));
+  uint64_t seed_unpacked = unpack_int_pyobj(seed);
+  // See Note [Acquire lock when using random generators]
+  std::lock_guard<std::mutex> lock(generator.mutex());
   generator.set_current_seed(seed_unpacked);
   Py_INCREF(self);
-  return (PyObject*)self;
+  return _self;
+  END_HANDLE_TH_ERRORS
+}
+
+static PyObject* THPGenerator_setOffset(PyObject* _self, PyObject* offset) {
+  HANDLE_TH_ERRORS
+  auto self = (THPGenerator*)_self;
+  auto generator = self->cdata;
+  THPUtils_assert(
+      THPUtils_checkLong(offset),
+      "manual_offset expected a long, "
+      "but got %s",
+      THPUtils_typename(offset));
+  uint64_t offset_unpacked = unpack_int_pyobj(offset);
+  // See Note [Acquire lock when using random generators]
+  std::lock_guard<std::mutex> lock(generator.mutex());
+  generator.set_offset(offset_unpacked);
+  Py_INCREF(self);
```

Comment:
1) Incref where you get the object, not where you return it, so move this up to `auto self`
2) This returns a borrowed ref to _self, as it has an incref. Where is it cleared? It is a little odd to me that setting the offset incurs a reference count. I know THPGenerator_setState does this too, but I don't get it. Can you explain the lifecycle here please?

Reply:
@albanD helped us understand this. Summary
* Arguments are always borrowed. So no need to incref them.
* CPython calling convention says that we need to send new references. Therefore we incref the new reference just before the return. Now CPython owns this reference, and will call DECREF when the new reference goes out of scope.